# Regression Week 5: Feature Selection and LASSO (Interpretation)

In this notebook, you will use LASSO to select features, building on a pre-implemented solver for LASSO (using GraphLab Create, though you can use other solvers). You will:
* Run LASSO with different L1 penalties.
* Choose best L1 penalty using a validation set.
* Choose best L1 penalty using a validation set, with additional constraint on the size of subset.

In the second notebook, you will implement your own LASSO solver, using coordinate descent. 

# Fire up graphlab create

In [1]:
import graphlab

# Load in house sales data

Dataset is from house sales in King County, the region where the city of Seattle, WA is located.

In [2]:
sales = graphlab.SFrame('kc_house_data.gl/')

This non-commercial license of GraphLab Create is assigned to jmfansi@gmail.com and will expire on April 03, 2017. For commercial licensing options, visit https://dato.com/buy/.


[INFO] graphlab.cython.cy_server: GraphLab Create v1.10.1 started. Logging: /tmp/graphlab_server_1465652986.log


# Create new features

As in Week 2, we consider features that are some transformations of inputs.

In [3]:
from math import log, sqrt
sales['sqft_living_sqrt'] = sales['sqft_living'].apply(sqrt)
sales['sqft_lot_sqrt'] = sales['sqft_lot'].apply(sqrt)
sales['bedrooms_square'] = sales['bedrooms']*sales['bedrooms']

# In the dataset, 'floors' was defined with type string, 
# so we'll convert them to float, before creating a new feature.
sales['floors'] = sales['floors'].astype(float) 
sales['floors_square'] = sales['floors']*sales['floors']

* Squaring bedrooms will increase the separation between not many bedrooms (e.g. 1) and lots of bedrooms (e.g. 4) since 1^2 = 1 but 4^2 = 16. Consequently this variable will mostly affect houses with many bedrooms.
* On the other hand, taking square root of sqft_living will decrease the separation between big house and small house. The owner may not be exactly twice as happy for getting a house that is twice as big.

# Learn regression weights with L1 penalty

Let us fit a model with all the features available, plus the features we just created above.

In [4]:
all_features = ['bedrooms', 'bedrooms_square',
            'bathrooms',
            'sqft_living', 'sqft_living_sqrt',
            'sqft_lot', 'sqft_lot_sqrt',
            'floors', 'floors_square',
            'waterfront', 'view', 'condition', 'grade',
            'sqft_above',
            'sqft_basement',
            'yr_built', 'yr_renovated']

Applying L1 penalty requires adding an extra parameter (`l1_penalty`) to the linear regression call in GraphLab Create. (Other tools may have separate implementations of LASSO.)  Note that it's important to set `l2_penalty=0` to ensure we don't introduce an additional L2 penalty.

In [5]:
model_all = graphlab.linear_regression.create(sales, target='price', features=all_features,
                                              validation_set=None, 
                                              l2_penalty=0., l1_penalty=1e10)

Linear regression:

--------------------------------------------------------

Number of examples          : 21613

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000002  | 1.426288     | 6962915.603493     | 426631.749026 |

| 2         | 3        | 0.000002  | 1.476852     | 6843144.200219     | 392488.929838 |

| 3         | 4        | 0.000002  | 1.522724     | 6831900.032123     | 385340.166783 |

| 4         | 5        | 0.000002  | 1.580564     | 6847166.848958     | 384842.383767 |

| 5         | 6        | 0.000002  | 1.630792     | 6869667.895833     | 385998.458623 |

| 6         | 7        | 0.000002  | 1.709357     | 6847177.773672     | 380824.455891 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Find what features had non-zero weight.

In [28]:
all_coefficients = model_all.get('coefficients')
non_ero_weight_features = all_coefficients[all_coefficients['value']!=0.0]['name']
print 'Non_Zero_Weight (%d)' % non_ero_weight_features.size()
print non_ero_weight_features
zer_weight_features = all_coefficients[all_coefficients['value']==0.0]['name']
print 'Zero_Weight (%d)' % zer_weight_features.size()
print zer_weight_features

Non_Zero_Weight (6)
['(intercept)', 'bathrooms', 'sqft_living', 'sqft_living_sqrt', 'grade', 'sqft_above']
Zero_Weight (12)
['bedrooms', 'bedrooms_square', 'sqft_lot', 'sqft_lot_sqrt', 'floors', 'floors_square', 'waterfront', 'view', 'condition', 'sqft_basement', 'yr_built', 'yr_renovated']


Note that a majority of the weights have been set to zero. So by setting an L1 penalty that's large enough, we are performing a subset selection. 

***QUIZ QUESTION***:
According to this list of weights, which of the features have been chosen? 

ANS:'bathrooms', 'sqft_living', 'sqft_living_sqrt', 'grade', 'sqft_above'

# Selecting an L1 penalty

To find a good L1 penalty, we will explore multiple values using a validation set. Let us do three way split into train, validation, and test sets:
* Split our sales data into 2 sets: training and test
* Further split our training data into two sets: train, validation

Be *very* careful that you use seed = 1 to ensure you get the same answer!

In [29]:
(training_and_validation, testing) = sales.random_split(.9,seed=1) # initial train/test split
(training, validation) = training_and_validation.random_split(0.5, seed=1) # split training into train and validate

Next, we write a loop that does the following:
* For `l1_penalty` in [10^1, 10^1.5, 10^2, 10^2.5, ..., 10^7] (to get this in Python, type `np.logspace(1, 7, num=13)`.)
    * Fit a regression model with a given `l1_penalty` on TRAIN data. Specify `l1_penalty=l1_penalty` and `l2_penalty=0.` in the parameter list.
    * Compute the RSS on VALIDATION data (here you will want to use `.predict()`) for that `l1_penalty`
* Report which `l1_penalty` produced the lowest RSS on validation data.

When you call `linear_regression.create()` make sure you set `validation_set = None`.

Note: you can turn off the print out of `linear_regression.create()` with `verbose = False`

In [32]:
import numpy as np
min_errors = None
best_l1 = None
for l1_penalty in np.logspace(1, 7, num=13):
    #Fitting the model
    model = graphlab.linear_regression.create(training, target='price', features=all_features,
                                              validation_set=None, 
                                              l2_penalty=0., l1_penalty=l1_penalty)
    #computing the RSS
    predictions = model.predict(validation)
    errors = validation['price']-predictions
    RSS = np.dot(errors, errors)
    if(min_errors == None or RSS < min_errors):
        min_errors = RSS
        best_l1 = l1_penalty
print 'The Best L1 Penalty is %f. This produces an RSS of %f' % (best_l1, min_errors)

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.345537     | 6287414.310279     | 322429.554621 |

| 2         | 3        | 0.000003  | 0.370332     | 6004129.870008     | 303352.364502 |

| 3         | 4        | 0.000003  | 0.426162     | 5888882.499599     | 298370.510858 |

| 4         | 5        | 0.000003  | 0.451171     | 5813781.207975     | 293774.670687 |

| 5         | 6        | 0.000003  | 0.478282     | 5737572.318059     | 289374.784200 |

| 6         | 7        | 0.000003  | 0.502115     | 5655864.782013     | 285354.826809 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.374968     | 6287414.313562     | 322429.554953 |

| 2         | 3        | 0.000003  | 0.407055     | 6004129.874269     | 303352.364626 |

| 3         | 4        | 0.000003  | 0.430595     | 5888882.504575     | 298370.510940 |

| 4         | 5        | 0.000003  | 0.455713     | 5813781.213635     | 293774.670786 |

| 5         | 6        | 0.000003  | 0.479903     | 5737572.324432     | 289374.784313 |

| 6         | 7        | 0.000003  | 0.501021     | 5655864.789116     | 285354.826924 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.367223     | 6287414.323944     | 322429.556003 |

| 2         | 3        | 0.000003  | 0.421616     | 6004129.887744     | 303352.365020 |

| 3         | 4        | 0.000003  | 0.475248     | 5888882.520311     | 298370.511200 |

| 4         | 5        | 0.000003  | 0.514060     | 5813781.231534     | 293774.671099 |

| 5         | 6        | 0.000003  | 0.575922     | 5737572.344586     | 289374.784672 |

| 6         | 7        | 0.000003  | 0.621944     | 5655864.811580     | 285354.827288 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.281793     | 6287414.356773     | 322429.559323 |

| 2         | 3        | 0.000003  | 0.302577     | 6004129.930357     | 303352.366264 |

| 3         | 4        | 0.000003  | 0.337894     | 5888882.570072     | 298370.512023 |

| 4         | 5        | 0.000003  | 0.381477     | 5813781.288134     | 293774.672088 |

| 5         | 6        | 0.000003  | 0.407161     | 5737572.408316     | 289374.785805 |

| 6         | 7        | 0.000003  | 0.439010     | 5655864.882615     | 285354.828441 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.329355     | 6287414.460587     | 322429.569820 |

| 2         | 3        | 0.000003  | 0.353500     | 6004130.065110     | 303352.370200 |

| 3         | 4        | 0.000003  | 0.395773     | 5888882.727430     | 298370.514625 |

| 4         | 5        | 0.000003  | 0.435346     | 5813781.467118     | 293774.675217 |

| 5         | 6        | 0.000003  | 0.484484     | 5737572.609850     | 289374.789391 |

| 6         | 7        | 0.000003  | 0.518179     | 5655865.107248     | 285354.832084 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.350398     | 6287414.788876     | 322429.603015 |

| 2         | 3        | 0.000003  | 0.425870     | 6004130.491238     | 303352.382646 |

| 3         | 4        | 0.000003  | 0.492040     | 5888883.225042     | 298370.522854 |

| 4         | 5        | 0.000003  | 0.514848     | 5813782.033117     | 293774.685112 |

| 5         | 6        | 0.000003  | 0.559862     | 5737573.247157     | 289374.800728 |

| 6         | 7        | 0.000003  | 0.611136     | 5655865.817602     | 285354.843605 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.381150     | 6287415.827019     | 322429.707988 |

| 2         | 3        | 0.000003  | 0.417436     | 6004131.838771     | 303352.422004 |

| 3         | 4        | 0.000003  | 0.448632     | 5888884.798629     | 298370.548875 |

| 4         | 5        | 0.000003  | 0.480385     | 5813783.822962     | 293774.716403 |

| 5         | 6        | 0.000003  | 0.510606     | 5737575.262497     | 289374.836581 |

| 6         | 7        | 0.000003  | 0.557684     | 5655868.063937     | 285354.880039 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.303799     | 6287419.109914     | 322430.039943 |

| 2         | 3        | 0.000003  | 0.334813     | 6004136.100045     | 303352.546466 |

| 3         | 4        | 0.000003  | 0.357180     | 5888889.774746     | 298370.631163 |

| 4         | 5        | 0.000003  | 0.394569     | 5813789.482948     | 293774.815357 |

| 5         | 6        | 0.000003  | 0.433278     | 5737581.635561     | 289374.949962 |

| 6         | 7        | 0.000003  | 0.467506     | 5655875.167473     | 285354.995257 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.494909     | 6287429.491340     | 322431.089695 |

| 2         | 3        | 0.000003  | 0.539288     | 6004149.575378     | 303352.940082 |

| 3         | 4        | 0.000003  | 0.561109     | 5888905.510609     | 298370.891417 |

| 4         | 5        | 0.000003  | 0.588907     | 5813807.381396     | 293775.128315 |

| 5         | 6        | 0.000003  | 0.649172     | 5737601.788960     | 289375.308545 |

| 6         | 7        | 0.000003  | 0.676830     | 5655897.630825     | 285355.359658 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.326667     | 6287462.320295     | 322434.409477 |

| 2         | 3        | 0.000003  | 0.346198     | 6004192.188124     | 303354.185095 |

| 3         | 4        | 0.000003  | 0.380471     | 5888955.271783     | 298371.714747 |

| 4         | 5        | 0.000003  | 0.403832     | 5813863.981264     | 293776.118347 |

| 5         | 6        | 0.000003  | 0.434826     | 5737665.519606     | 289376.442902 |

| 6         | 7        | 0.000003  | 0.475415     | 5655968.666186     | 285356.512471 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.392962     | 6287566.134553     | 322444.909304 |

| 2         | 3        | 0.000003  | 0.430045     | 6004326.941447     | 303358.125094 |

| 3         | 4        | 0.000003  | 0.469285     | 5889112.630416     | 298374.321707 |

| 4         | 5        | 0.000003  | 0.495797     | 5814042.965743     | 293779.252821 |

| 5         | 6        | 0.000003  | 0.551395     | 5737867.053589     | 289380.034238 |

| 6         | 7        | 0.000003  | 0.592891     | 5656193.299699     | 285360.162764 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.378575     | 6287894.424085     | 322478.130206 |

| 2         | 3        | 0.000003  | 0.417939     | 6004753.068898     | 303370.613672 |

| 3         | 4        | 0.000003  | 0.449250     | 5889610.242137     | 298382.599251 |

| 4         | 5        | 0.000003  | 0.477221     | 5814608.964400     | 293789.202075 |

| 5         | 6        | 0.000003  | 0.504690     | 5738504.360039     | 289391.432870 |

| 6         | 7        | 0.000003  | 0.529969     | 5656903.653287     | 285371.753783 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.276670     | 6288932.566682     | 322583.359172 |

| 2         | 3        | 0.000003  | 0.316689     | 6006100.602140     | 303410.398042 |

| 3         | 4        | 0.000003  | 0.419432     | 5891183.828480     | 298409.111215 |

| 4         | 5        | 0.000003  | 0.445395     | 5816398.809211     | 293821.036075 |

| 5         | 6        | 0.000003  | 0.505869     | 5740519.699879     | 289427.896727 |

| 6         | 7        | 0.000003  | 0.542946     | 5659149.988446     | 285408.885485 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

The Best L1 Penalty is 10.000000. This produces an RSS of 625766285142460.500000


In [33]:
best_model = graphlab.linear_regression.create(training, target='price', features=all_features,
                                              validation_set=None, 
                                              l2_penalty=0., l1_penalty=10.0)
#computing the RSS
predictions = best_model.predict(testing)
errors = testing['price']-predictions
RSS = np.dot(errors, errors)

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.288259     | 6287414.310279     | 322429.554621 |

| 2         | 3        | 0.000003  | 0.366744     | 6004129.870008     | 303352.364502 |

| 3         | 4        | 0.000003  | 0.389116     | 5888882.499599     | 298370.510858 |

| 4         | 5        | 0.000003  | 0.423556     | 5813781.207975     | 293774.670687 |

| 5         | 6        | 0.000003  | 0.464557     | 5737572.318059     | 289374.784200 |

| 6         | 7        | 0.000003  | 0.497439     | 5655864.782013     | 285354.826809 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

*** QUIZ QUESTIONS ***
1. What was the best value for the `l1_penalty`?

ANS:10.000000
2. What is the RSS on TEST data of the model with the best `l1_penalty`?

ANS: 1.56983602382e+14


In [34]:
print RSS

1.56983602382e+14


***QUIZ QUESTION***
Also, using this value of L1 penalty, how many nonzero weights do you have?

ANS: 18

In [38]:
best_model_coeff = best_model.get('coefficients')
non_ero_weight_features_2 = best_model_coeff[best_model_coeff['value']!=0.0]['name']
print 'there are %d non zeros features' % (non_ero_weight_features_2.size())
print non_ero_weight_features_2

there are 18 non zeros features
['(intercept)', 'bedrooms', 'bedrooms_square', 'bathrooms', 'sqft_living', 'sqft_living_sqrt', 'sqft_lot', 'sqft_lot_sqrt', 'floors', 'floors_square', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated']
18


# Limit the number of nonzero weights

What if we absolutely wanted to limit ourselves to, say, 7 features? This may be important if we want to derive "a rule of thumb" --- an interpretable model that has only a few features in them.

In this section, you are going to implement a simple, two phase procedure to achive this goal:
1. Explore a large range of `l1_penalty` values to find a narrow region of `l1_penalty` values where models are likely to have the desired number of non-zero weights.
2. Further explore the narrow region you found to find a good value for `l1_penalty` that achieves the desired sparsity.  Here, we will again use a validation set to choose the best value for `l1_penalty`.

In [36]:
max_nonzeros = 7

## Exploring the larger range of values to find a narrow range with the desired sparsity

Let's define a wide range of possible `l1_penalty_values`:

In [65]:
l1_penalty_values = np.logspace(8, 10, num=20)

Now, implement a loop that search through this space of possible `l1_penalty` values:

* For `l1_penalty` in `np.logspace(8, 10, num=20)`:
    * Fit a regression model with a given `l1_penalty` on TRAIN data. Specify `l1_penalty=l1_penalty` and `l2_penalty=0.` in the parameter list. When you call `linear_regression.create()` make sure you set `validation_set = None`
    * Extract the weights of the model and count the number of nonzeros. Save the number of nonzeros to a list.
        * *Hint: `model['coefficients']['value']` gives you an SArray with the parameters you learned.  If you call the method `.nnz()` on it, you will find the number of non-zero parameters!* 

In [66]:
non_zeroes = []
for l1_penalty in np.logspace(8, 10, num=20):
    #Fitting the model
    model = graphlab.linear_regression.create(training, target='price', features=all_features,
                                              validation_set=None, 
                                              l2_penalty=0., l1_penalty=l1_penalty)
    non_zeroes.append(model['coefficients']['value'].nnz())

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.161879     | 6302596.887972     | 323993.114447 |

| 2         | 3        | 0.000003  | 0.177273     | 6023837.209073     | 303975.326597 |

| 3         | 4        | 0.000003  | 0.192065     | 5911895.809122     | 298805.600064 |

| 4         | 5        | 0.000003  | 0.207296     | 5839957.243889     | 294292.598204 |

| 5         | 6        | 0.000003  | 0.222049     | 5767046.162784     | 289966.959726 |

| 6         | 7        | 0.000003  | 0.236029     | 5688716.875912     | 285965.139010 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.296424     | 6306761.089277     | 324431.811110 |

| 2         | 3        | 0.000003  | 0.313818     | 6029242.439247     | 304162.702383 |

| 3         | 4        | 0.000003  | 0.332655     | 5918207.784303     | 298943.962355 |

| 4         | 5        | 0.000003  | 0.362562     | 5847136.675810     | 294455.684490 |

| 5         | 6        | 0.000003  | 0.394418     | 5775130.101176     | 290153.028407 |

| 6         | 7        | 0.000003  | 0.435764     | 5697727.383553     | 286159.539614 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.338298     | 6312067.428647     | 324996.927681 |

| 2         | 3        | 0.000003  | 0.355829     | 6036130.191201     | 304411.734396 |

| 3         | 4        | 0.000003  | 0.369555     | 5926250.979130     | 299132.112018 |

| 4         | 5        | 0.000003  | 0.415653     | 5856285.249441     | 294676.578872 |

| 5         | 6        | 0.000003  | 0.441114     | 5785431.265169     | 290404.829180 |

| 6         | 7        | 0.000003  | 0.500093     | 5709209.251967     | 286424.040727 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.266247     | 6318829.161558     | 325726.887556 |

| 2         | 3        | 0.000003  | 0.284487     | 6044907.077830     | 304745.693519 |

| 3         | 4        | 0.000003  | 0.303841     | 5936500.217141     | 299391.051314 |

| 4         | 5        | 0.000003  | 0.333965     | 5867943.043467     | 294979.244899 |

| 5         | 6        | 0.000003  | 0.357954     | 5798557.775695     | 290749.495050 |

| 6         | 7        | 0.000003  | 0.375717     | 5723840.304020     | 286788.259115 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.280852     | 6327445.469597     | 326672.935878 |

| 2         | 3        | 0.000003  | 0.309650     | 6056091.246122     | 305198.151982 |

| 3         | 4        | 0.000003  | 0.363259     | 5949560.566172     | 299752.080188 |

| 4         | 5        | 0.000003  | 0.381923     | 5882798.280387     | 295399.217084 |

| 5         | 6        | 0.000003  | 0.425143     | 5815284.561515     | 291227.202999 |

| 6         | 7        | 0.000003  | 0.457029     | 5742484.289679     | 287296.320924 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.301023     | 6338425.015323     | 327904.016823 |

| 2         | 3        | 0.000003  | 0.342559     | 6070342.951911     | 305818.193076 |

| 3         | 4        | 0.000003  | 0.377914     | 5966203.042130     | 300262.391571 |

| 4         | 5        | 0.000003  | 0.412010     | 5901727.937067     | 295989.815933 |

| 5         | 6        | 0.000003  | 0.462270     | 5836599.086125     | 291898.147485 |

| 6         | 7        | 0.000003  | 0.489593     | 5766241.854070     | 288014.702465 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.268089     | 6352415.978265     | 329513.773176 |

| 2         | 3        | 0.000003  | 0.289421     | 6088503.547309     | 306678.464467 |

| 3         | 4        | 0.000003  | 0.311295     | 5987410.136686     | 300993.852916 |

| 4         | 5        | 0.000003  | 0.337283     | 5925849.529320     | 296831.866711 |

| 5         | 6        | 0.000003  | 0.362430     | 5863759.656463     | 292853.418793 |

| 6         | 7        | 0.000003  | 0.382330     | 5796515.529440     | 289044.471466 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.275665     | 6370244.310167     | 331630.660676 |

| 2         | 3        | 0.000003  | 0.298162     | 6111645.136891     | 307887.639691 |

| 3         | 4        | 0.000003  | 0.341744     | 6014433.803461     | 302056.769305 |

| 4         | 5        | 0.000003  | 0.391847     | 5956587.067149     | 298048.868696 |

| 5         | 6        | 0.000003  | 0.419541     | 5898369.687568     | 294231.926395 |

| 6         | 7        | 0.000003  | 0.451601     | 5835068.031486     | 290539.467795 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.393734     | 6392962.501428     | 334432.629813 |

| 2         | 3        | 0.000003  | 0.425115     | 6141133.877684     | 309609.652192 |

| 3         | 4        | 0.000003  | 0.448149     | 6048869.376781     | 303621.347701 |

| 4         | 5        | 0.000003  | 0.487204     | 5995755.132177     | 299830.533427 |

| 5         | 6        | 0.000003  | 0.512304     | 5942472.372582     | 296246.537804 |

| 6         | 7        | 0.000003  | 0.546023     | 5882772.472167     | 292655.514067 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.370993     | 6421911.739799     | 338168.430249 |

| 2         | 3        | 0.000003  | 0.400893     | 6178710.662590     | 312093.147503 |

| 3         | 4        | 0.000003  | 0.426967     | 6092749.790017     | 305950.822078 |

| 4         | 5        | 0.000003  | 0.460122     | 6045666.044451     | 302468.872046 |

| 5         | 6        | 0.000003  | 0.479945     | 5998277.020726     | 299195.437483 |

| 6         | 7        | 0.000003  | 0.501101     | 5944444.613403     | 295775.972512 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.238000     | 6458801.015264     | 343188.313914 |

| 2         | 3        | 0.000003  | 0.256294     | 6226593.800368     | 315715.881133 |

| 3         | 4        | 0.000003  | 0.278230     | 6148665.480461     | 309451.624488 |

| 4         | 5        | 0.000003  | 0.299617     | 6109266.245878     | 306412.359795 |

| 5         | 6        | 0.000003  | 0.319877     | 6067941.501592     | 303417.317296 |

| 6         | 7        | 0.000003  | 0.350297     | 6019163.317913     | 300393.085687 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.287561     | 6505808.107267     | 349987.699359 |

| 2         | 3        | 0.000003  | 0.330952     | 6287610.099403     | 321049.939717 |

| 3         | 4        | 0.000003  | 0.383346     | 6219917.462832     | 314747.103160 |

| 4         | 5        | 0.000003  | 0.415381     | 6186864.205766     | 312133.200443 |

| 5         | 6        | 0.000003  | 0.457392     | 6141049.702542     | 309268.413008 |

| 6         | 7        | 0.000003  | 0.499955     | 6098090.379557     | 306911.599029 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.190478     | 6565708.060916     | 359267.674436 |

| 2         | 3        | 0.000003  | 0.209376     | 6365361.632804     | 328953.510730 |

| 3         | 4        | 0.000003  | 0.225156     | 6305829.246141     | 322613.170450 |

| 4         | 5        | 0.000003  | 0.292461     | 6265428.838674     | 319852.473118 |

| 5         | 6        | 0.000003  | 0.311421     | 6233853.106449     | 318320.600901 |

| 6         | 7        | 0.000003  | 0.346296     | 6199198.857716     | 316732.370565 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.340884     | 6642037.071382     | 372015.681824 |

| 2         | 3        | 0.000003  | 0.370381     | 6461738.956416     | 340588.423328 |

| 3         | 4        | 0.000003  | 0.404746     | 6398392.723924     | 333431.863633 |

| 4         | 5        | 0.000003  | 0.427954     | 6371877.122886     | 331286.359425 |

| 5         | 6        | 0.000003  | 0.455370     | 6353136.654069     | 330421.269864 |

| 6         | 7        | 0.000003  | 0.475501     | 6330612.934807     | 329296.764297 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.300835     | 6739301.231007     | 389607.085142 |

| 2         | 3        | 0.000003  | 0.339005     | 6569297.382654     | 356355.616933 |

| 3         | 4        | 0.000003  | 0.367285     | 6519180.189446     | 348232.602946 |

| 4         | 5        | 0.000003  | 0.401627     | 6511771.934086     | 347519.257327 |

| 5         | 6        | 0.000003  | 0.434841     | 6508236.398022     | 347641.774942 |

| 6         | 7        | 0.000003  | 0.457231     | 6496515.714757     | 346711.084406 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.291401     | 6851251.348758     | 412576.866347 |

| 2         | 3        | 0.000003  | 0.316406     | 6705927.418436     | 378624.562082 |

| 3         | 4        | 0.000003  | 0.339531     | 6678759.248861     | 371401.977561 |

| 4         | 5        | 0.000003  | 0.372324     | 6689839.793283     | 371772.510021 |

| 5         | 6        | 0.000003  | 0.402343     | 6700050.222874     | 372113.713871 |

| 6         | 7        | 0.000003  | 0.435916     | 6697293.370032     | 370685.051067 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.320795     | 6990727.771524     | 443765.694135 |

| 2         | 3        | 0.000003  | 0.341103     | 6881811.580352     | 411414.484829 |

| 3         | 4        | 0.000003  | 0.363584     | 6872119.637764     | 404347.406375 |

| 4         | 5        | 0.000003  | 0.387246     | 6884564.579949     | 402804.042129 |

| 5         | 6        | 0.000003  | 0.423198     | 6884279.257633     | 400745.316819 |

| 6         | 7        | 0.000003  | 0.442864     | 6869773.133183     | 397237.306904 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.349610     | 7165064.642471     | 485718.879258 |

| 2         | 3        | 0.000003  | 0.371654     | 7086725.698544     | 456087.127286 |

| 3         | 4        | 0.000003  | 0.392086     | 7075577.209652     | 446853.928812 |

| 4         | 5        | 0.000003  | 0.431270     | 7062536.185828     | 441456.188174 |

| 5         | 6        | 0.000003  | 0.453160     | 7054349.178409     | 436331.029744 |

| 6         | 7        | 0.000003  | 0.469597     | 7053909.066828     | 430723.907200 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000005  | 0.260453     | 7202439.442101     | 484432.625377 |

| 2         | 3        | 0.000005  | 0.290287     | 7284840.298565     | 505076.954827 |

| 3         | 4        | 0.000005  | 0.325347     | 7266438.445660     | 490990.237855 |

| 4         | 5        | 0.000005  | 0.363313     | 7270633.139688     | 473802.625159 |

| 5         | 6        | 0.000005  | 0.388414     | 7320904.689497     | 462020.358653 |

| 6         | 7        | 0.000005  | 0.407655     | 7344719.665332     | 441190.123985 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000012  | 0.235772     | 7228843.829153     | 476257.577720 |

| 2         | 3        | 0.000012  | 0.258796     | 7504641.811821     | 514487.205685 |

| 3         | 4        | 0.000012  | 0.283756     | 7402647.878877     | 451928.315910 |

| 4         | 5        | 0.000012  | 0.307441     | 7312309.081776     | 409957.463247 |

| 5         | 6        | 0.000012  | 0.339155     | 7239807.218168     | 388380.843122 |

| 6         | 7        | 0.000012  | 0.359607     | 7187431.218291     | 380705.223778 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Out of this large range, we want to find the two ends of our desired narrow range of `l1_penalty`.  At one end, we will have `l1_penalty` values that have too few non-zeros, and at the other end, we will have an `l1_penalty` that has too many non-zeros.  

More formally, find:
* The largest `l1_penalty` that has more non-zeros than `max_nonzero` (if we pick a penalty smaller than this value, we will definitely have too many non-zero weights)
    * Store this value in the variable `l1_penalty_min` (we will use it later)
* The smallest `l1_penalty` that has fewer non-zeros than `max_nonzero` (if we pick a penalty larger than this value, we will definitely have too few non-zero weights)
    * Store this value in the variable `l1_penalty_max` (we will use it later)


*Hint: there are many ways to do this, e.g.:*
* Programmatically within the loop above
* Creating a list with the number of non-zeros for each value of `l1_penalty` and inspecting it to find the appropriate boundaries.

In [70]:
l1_penalty =  np.logspace(8, 10, num=20)
largest_non_zero_index=0
largest_l1_penalty_with_largest_non_zero=None
smallest_l1_penalty_with_smallest_non_zero=None
index=0
for nz in non_zeroes:   
    if nz > max_nonzeros: 
        current_max_l1 = l1_penalty[index]
        if(largest_l1_penalty_with_largest_non_zero == None or current_max_l1 > largest_l1_penalty_with_largest_non_zero):
            largest_l1_penalty_with_largest_non_zero = current_max_l1
    elif nz < max_nonzeros:
        current_min_l1 = l1_penalty[index]
        if(smallest_l1_penalty_with_smallest_non_zero == None or current_min_l1 < smallest_l1_penalty_with_smallest_non_zero):
            smallest_l1_penalty_with_smallest_non_zero=current_min_l1
    index = index +1

l1_penalty_min = largest_l1_penalty_with_largest_non_zero
l1_penalty_max = smallest_l1_penalty_with_smallest_non_zero
print 'l1 penalty min is %f' % (l1_penalty_min)
print 'l1 penalty max is %f' % (l1_penalty_max)


l1 penalty min is 2976351441.631313
l1 penalty max is 3792690190.732254


***QUIZ QUESTIONS***

What values did you find for `l1_penalty_min` and`l1_penalty_max`? 

ANS: (l1_penalty_min, l1_penalty_max) = (2976351441.631313, 3792690190.732254)

## Exploring the narrow range of values to find the solution with the right number of non-zeros that has lowest RSS on the validation set 

We will now explore the narrow region of `l1_penalty` values we found:

In [71]:
l1_penalty_values = np.linspace(l1_penalty_min,l1_penalty_max,20)

* For `l1_penalty` in `np.linspace(l1_penalty_min,l1_penalty_max,20)`:
    * Fit a regression model with a given `l1_penalty` on TRAIN data. Specify `l1_penalty=l1_penalty` and `l2_penalty=0.` in the parameter list. When you call `linear_regression.create()` make sure you set `validation_set = None`
    * Measure the RSS of the learned model on the VALIDATION set

Find the model that the lowest RSS on the VALIDATION set and has sparsity *equal* to `max_nonzero`.

In [76]:
min_rss = None
select_l1= None
selectedModel = None
for l1_penalty in l1_penalty_values:
    model = graphlab.linear_regression.create(training, target='price', features=all_features,
                                              validation_set=None, 
                                             l2_penalty=0., l1_penalty=l1_penalty)
    
    if (model['coefficients']['value'].nnz()-1 == max_nonzeros):
        #computing the RSS
        predictions = model.predict(validation)
        errors = validation['price']-predictions
        RSS = np.dot(errors, errors)
        if(min_rss==None or RSS < min_rss):
            min_rss=RSS
            select_l1 = l1_penalty
            selectedModel = model

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.189106     | 6739301.231007     | 389607.085142 |

| 2         | 3        | 0.000003  | 0.206296     | 6569297.382654     | 356355.616933 |

| 3         | 4        | 0.000003  | 0.229822     | 6519180.189446     | 348232.602946 |

| 4         | 5        | 0.000003  | 0.250024     | 6511771.934086     | 347519.257327 |

| 5         | 6        | 0.000003  | 0.274942     | 6508236.398022     | 347641.774942 |

| 6         | 7        | 0.000003  | 0.315262     | 6496515.714757     | 346711.084406 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.345938     | 6745824.463817     | 390816.789246 |

| 2         | 3        | 0.000003  | 0.371406     | 6576394.847047     | 357433.566852 |

| 3         | 4        | 0.000003  | 0.404515     | 6527572.935664     | 349357.366319 |

| 4         | 5        | 0.000003  | 0.471791     | 6521142.806632     | 348701.019139 |

| 5         | 6        | 0.000003  | 0.524946     | 6518614.014289     | 348888.199260 |

| 6         | 7        | 0.000003  | 0.564630     | 6507128.891315     | 347885.665556 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.305807     | 6752347.657761     | 392009.121753 |

| 2         | 3        | 0.000003  | 0.341867     | 6583521.070903     | 358524.867897 |

| 3         | 4        | 0.000003  | 0.360167     | 6535971.427822     | 350493.642380 |

| 4         | 5        | 0.000003  | 0.379775     | 6530517.602470     | 349894.204388 |

| 5         | 6        | 0.000003  | 0.397014     | 6528996.173933     | 350146.209214 |

| 6         | 7        | 0.000003  | 0.415749     | 6517746.461337     | 349069.317383 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.262853     | 6758338.328596     | 393183.638571 |

| 2         | 3        | 0.000003  | 0.295932     | 6590703.295167     | 359634.836047 |

| 3         | 4        | 0.000003  | 0.316229     | 6544370.536454     | 351640.727736 |

| 4         | 5        | 0.000003  | 0.336332     | 6539890.228171     | 351097.892318 |

| 5         | 6        | 0.000003  | 0.357399     | 6539379.516775     | 351415.297011 |

| 6         | 7        | 0.000003  | 0.375096     | 6528366.248329     | 350261.721339 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.262379     | 6764145.398843     | 394355.637359 |

| 2         | 3        | 0.000003  | 0.305133     | 6597904.810916     | 360757.199941 |

| 3         | 4        | 0.000003  | 0.332011     | 6552769.840363     | 352798.630865 |

| 4         | 5        | 0.000003  | 0.360050     | 6549262.086461     | 352312.355494 |

| 5         | 6        | 0.000003  | 0.392694     | 6549763.246198     | 352695.353562 |

| 6         | 7        | 0.000003  | 0.416168     | 6538986.777810     | 351462.667919 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.295262     | 6769952.469091     | 395533.206039 |

| 2         | 3        | 0.000003  | 0.320968     | 6605106.326666     | 361888.548761 |

| 3         | 4        | 0.000003  | 0.347166     | 6561169.144272     | 353967.313247 |

| 4         | 5        | 0.000003  | 0.369107     | 6558633.944750     | 353537.695247 |

| 5         | 6        | 0.000003  | 0.392906     | 6560146.975621     | 353986.270792 |

| 6         | 7        | 0.000003  | 0.435137     | 6549607.307292     | 352672.009019 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.284111     | 6775759.539339     | 396716.295011 |

| 2         | 3        | 0.000003  | 0.326567     | 6612307.842416     | 363028.798505 |

| 3         | 4        | 0.000003  | 0.346992     | 6569568.448181     | 355146.668469 |

| 4         | 5        | 0.000003  | 0.383596     | 6568005.803040     | 354773.798877 |

| 5         | 6        | 0.000003  | 0.413000     | 6570530.705045     | 355287.930317 |

| 6         | 7        | 0.000003  | 0.431354     | 6560227.836773     | 353889.658579 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.290983     | 6781566.574986     | 397904.847937 |

| 2         | 3        | 0.000003  | 0.319586     | 6619509.315257     | 364177.858692 |

| 3         | 4        | 0.000003  | 0.347256     | 6577967.702044     | 356336.583438 |

| 4         | 5        | 0.000003  | 0.407366     | 6577377.605489     | 356020.546809 |

| 5         | 6        | 0.000003  | 0.437391     | 6580914.372598     | 356600.206652 |

| 6         | 7        | 0.000003  | 0.467821     | 6570848.302974     | 355115.523804 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.299203     | 6787373.645234     | 399098.830100 |

| 2         | 3        | 0.000003  | 0.327328     | 6626710.831006     | 365335.659819 |

| 3         | 4        | 0.000003  | 0.357929     | 6586367.005953     | 357536.966832 |

| 4         | 5        | 0.000003  | 0.384186     | 6586749.463778     | 357277.842388 |

| 5         | 6        | 0.000003  | 0.433381     | 6591298.102022     | 357922.998572 |

| 6         | 7        | 0.000003  | 0.456544     | 6581468.832455     | 356349.534440 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.328715     | 6793180.715481     | 400298.185918 |

| 2         | 3        | 0.000003  | 0.348472     | 6633912.346756     | 366502.112290 |

| 3         | 4        | 0.000003  | 0.368493     | 6594766.309862     | 358747.706580 |

| 4         | 5        | 0.000003  | 0.392766     | 6596121.322068     | 358545.567335 |

| 5         | 6        | 0.000003  | 0.418099     | 6601681.831445     | 359256.182222 |

| 6         | 7        | 0.000003  | 0.441446     | 6592089.361936     | 357591.598955 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.322930     | 6798987.785729     | 401502.867233 |

| 2         | 3        | 0.000003  | 0.344708     | 6641113.862506     | 367677.133766 |

| 3         | 4        | 0.000003  | 0.365877     | 6603165.613771     | 359968.698182 |

| 4         | 5        | 0.000003  | 0.401447     | 6605493.180357     | 359823.611417 |

| 5         | 6        | 0.000003  | 0.426084     | 6612065.560868     | 360599.642341 |

| 6         | 7        | 0.000003  | 0.448207     | 6602709.891417     | 358841.633718 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.300293     | 6804794.855977     | 402712.826253 |

| 2         | 3        | 0.000003  | 0.328475     | 6648315.378256     | 368860.642357 |

| 3         | 4        | 0.000003  | 0.357282     | 6611564.917680     | 361199.837673 |

| 4         | 5        | 0.000003  | 0.381804     | 6614865.038647     | 361111.865068 |

| 5         | 6        | 0.000003  | 0.414904     | 6622449.290291     | 361953.264500 |

| 6         | 7        | 0.000003  | 0.440830     | 6613330.420898     | 360099.555727 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.300212     | 6810601.891624     | 403928.008294 |

| 2         | 3        | 0.000003  | 0.334002     | 6655516.851097     | 370052.549506 |

| 3         | 4        | 0.000003  | 0.352134     | 6619964.171543     | 362441.014218 |

| 4         | 5        | 0.000003  | 0.376124     | 6624236.841096     | 362410.211648 |

| 5         | 6        | 0.000003  | 0.400107     | 6632388.356592     | 363238.633808 |

| 6         | 7        | 0.000003  | 0.426429     | 6624497.937274     | 361462.325873 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.281187     | 6816408.961872     | 405148.380776 |

| 2         | 3        | 0.000003  | 0.299397     | 6662718.366846     | 371252.788456 |

| 3         | 4        | 0.000003  | 0.326049     | 6628363.475452     | 363692.139769 |

| 4         | 5        | 0.000003  | 0.346430     | 6633608.699386     | 363718.558461 |

| 5         | 6        | 0.000003  | 0.370958     | 6642054.345717     | 364484.037875 |

| 6         | 7        | 0.000003  | 0.413423     | 6636001.595070     | 362894.485099 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.312629     | 6822216.032119     | 406373.889776 |

| 2         | 3        | 0.000003  | 0.324072     | 6669919.882596     | 372461.271652 |

| 3         | 4        | 0.000003  | 0.337521     | 6636762.779361     | 364953.104716 |

| 4         | 5        | 0.000003  | 0.350766     | 6642980.557675     | 365036.790356 |

| 5         | 6        | 0.000003  | 0.366193     | 6651720.334842     | 365737.059513 |

| 6         | 7        | 0.000003  | 0.380978     | 6647505.252865     | 364336.765726 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.272395     | 6828023.102367     | 407604.488964 |

| 2         | 3        | 0.000003  | 0.288133     | 6677121.398346     | 373677.919108 |

| 3         | 4        | 0.000003  | 0.305790     | 6645162.083270     | 366223.807423 |

| 4         | 5        | 0.000003  | 0.349761     | 6652352.415965     | 366364.800631 |

| 5         | 6        | 0.000003  | 0.378543     | 6661386.323967     | 366997.620697 |

| 6         | 7        | 0.000003  | 0.412069     | 6659008.910661     | 365789.048032 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.330813     | 6833830.172615     | 408840.132375 |

| 2         | 3        | 0.000003  | 0.351766     | 6684322.914096     | 374902.651339 |

| 3         | 4        | 0.000003  | 0.376595     | 6653561.387179     | 367504.146882 |

| 4         | 5        | 0.000003  | 0.404004     | 6661724.274254     | 367702.483336 |

| 5         | 6        | 0.000003  | 0.424262     | 6671052.313092     | 368265.644004 |

| 6         | 7        | 0.000003  | 0.453922     | 6670512.568457     | 367251.213363 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.310126     | 6839637.208262     | 410080.767006 |

| 2         | 3        | 0.000003  | 0.332806     | 6691524.386936     | 376135.382003 |

| 3         | 4        | 0.000003  | 0.379010     | 6661960.641043     | 368794.015010 |

| 4         | 5        | 0.000003  | 0.402234     | 6671096.076703     | 369049.725239 |

| 5         | 6        | 0.000003  | 0.420876     | 6680718.244624     | 369541.044998 |

| 6         | 7        | 0.000003  | 0.442239     | 6679542.716608     | 368401.661442 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.333151     | 6845444.278510     | 411326.362411 |

| 2         | 3        | 0.000003  | 0.363263     | 6698725.902686     | 377376.047334 |

| 3         | 4        | 0.000003  | 0.392449     | 6670359.944952     | 370093.327468 |

| 4         | 5        | 0.000003  | 0.412434     | 6680467.934993     | 370406.438000 |

| 5         | 6        | 0.000003  | 0.453202     | 6690384.233749     | 370823.762675 |

| 6         | 7        | 0.000003  | 0.474965     | 6688418.043320     | 369539.610571 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Linear regression:

--------------------------------------------------------

Number of examples          : 9761

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

+-----------+----------+-----------+--------------+--------------------+---------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+-----------+--------------+--------------------+---------------+

Tuning step size. First iteration could take longer than subsequent iterations.

| 1         | 2        | 0.000003  | 0.310089     | 6851251.348758     | 412576.866347 |

| 2         | 3        | 0.000003  | 0.337592     | 6705927.418436     | 378624.562082 |

| 3         | 4        | 0.000003  | 0.365096     | 6678759.248861     | 371401.977561 |

| 4         | 5        | 0.000003  | 0.388068     | 6689839.793283     | 371772.510021 |

| 5         | 6        | 0.000003  | 0.420567     | 6700050.222874     | 372113.713871 |

| 6         | 7        | 0.000003  | 0.449598     | 6697293.370032     | 370685.051067 |

+-----------+----------+-----------+--------------+--------------------+---------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

***QUIZ QUESTIONS***
1. What value of `l1_penalty` in our narrow range has the lowest RSS on the VALIDATION set and has sparsity *equal* to `max_nonzeros`?
2. What features in this model have non-zero coefficients?

In [82]:
print min_rss
print select_l1
all_coef = selectedModel['coefficients']
non_zero = all_coef[all_coef['value']!=0.0]['name']
print non_zero
print non_zero.size()

1.03461690923e+15
3320073020.2
['(intercept)', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_living_sqrt', 'floors', 'grade', 'sqft_above', ... ]
8
